Update channel_shuffle_op and transpose 2d to speed up ShuffleNet

xiaomengy opened the pull request:
Summary: Update channel_shuffle_op and transpose 2d to speed up ShuffleNet

Differential Revision: D8889361

BIT-silence commented:
@pytorchbot retest this please

houseroad commented on caffe2/utils/math_cpu.cc:
```diff
@@ -2568,24 +2620,53 @@ void TransposeCPUImpl(
-#ifdef CAFFE2_USE_HPTT
-  if (TryTransposeWithHPTT(ndim, dims, axes, X, Y)) {
+    float* Y) {
+  if (utils::IsIdentityPermutation(ndim, axes)) {
```
Add another check, if X == Y, just return?

BIT-silence replied:
Currently transpose cannot support in-place. X must not be Y.

houseroad replied:
I see, make sense.

BIT-silence commented:
@pytorchbot retest this please

houseroad reviewed (approved):
The logic LGTM. Please make sure the iOS part is well tested.

BIT-silence commented:
@pytorchbot retest this please

bddppq commented:
@pytorchbot retest this please

houseroad commented:
@pytorchbot retest this please

houseroad commented on caffe2/operators/channel_shuffle_op.cc:
```diff
@@ -1,23 +1,219 @@
+      }
+    } else {
+      for (int j = 0; j < K; ++j) {
+        vDSP_mmov(X + j * HxW, Y + j * G * HxW, HxW, G, k * HxW, HxW);
```
typo: k ==> K

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

houseroad commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorch retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

xiaomengy committed:
Update channel_shuffle_op and transpose 2d to speed up ShuffleNet (#9525)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9525

Update channel_shuffle_op and transpose 2d to speed up ShuffleNet

Reviewed By: houseroad

Differential Revision: D8889361

fbshipit-source-id: b5344955acb466e224e1c80cbaca721743c09e2f
